File: Src/java/engine/src/test/java/org/opencds/cqf/cql/engine/runtime/ValueTest.java
package org.opencds.cqf.cql.engine.runtime;

import static org.junit.jupiter.api.Assertions.*;

import java.math.BigDecimal;
import java.util.stream.Stream;
import org.junit.jupiter.api.Test;

class ValueTest {

    @Test
    void getCoarsestScale() {
        assertEquals(0, Value.getCoarsestScale(Stream.of()));
        assertEquals(2, Value.getCoarsestScale(Stream.of(new BigDecimal("1.1234"), new BigDecimal("1.12"), null)));
    }

    @Test
    void roundToScale() {
        assertEquals(new BigDecimal("1.12"), Value.roundToScale(new BigDecimal("1.1234"), 2, false));
        assertEquals(new BigDecimal("1.13"), Value.roundToScale(new BigDecimal("1.1234"), 2, true));
        assertEquals(new BigDecimal("1.12"), Value.roundToScale(new BigDecimal("1.1200"), 2, false));
        assertEquals(new BigDecimal("1.12"), Value.roundToScale(new BigDecimal("1.1200"), 2, true));
        assertEquals(new BigDecimal("1.12"), Value.roundToScale(new BigDecimal("1.12"), 4, false));
        assertEquals(new BigDecimal("1.12"), Value.roundToScale(new BigDecimal("1.12"), 4, true));
    }
}

File: Src/java/engine/src/test/kotlin/org/opencds/cqf/cql/engine/runtime/ValueTest.kt
package org.opencds.cqf.cql.engine.runtime

import java.math.BigDecimal
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

internal class ValueTest {
    @Test
    fun getCoarsestScale() {
        Assertions.assertEquals(0, Value.getCoarsestScale(listOf()))
        Assertions.assertEquals(
            2,
            Value.getCoarsestScale(listOf(BigDecimal("1.1234"), BigDecimal("1.12"), null)),
        )
    }

    @Test
    fun roundToScale() {
        Assertions.assertEquals(
            BigDecimal("1.12"),
            Value.roundToScale(BigDecimal("1.1234"), 2, false),
        )
        Assertions.assertEquals(
            BigDecimal("1.13"),
            Value.roundToScale(BigDecimal("1.1234"), 2, true),
        )
        Assertions.assertEquals(
            BigDecimal("1.12"),
            Value.roundToScale(BigDecimal("1.1200"), 2, false),
        )
        Assertions.assertEquals(
            BigDecimal("1.12"),
            Value.roundToScale(BigDecimal("1.1200"), 2, true),
        )
        Assertions.assertEquals(
            BigDecimal("1.12"),
            Value.roundToScale(BigDecimal("1.12"), 4, false),
        )
        Assertions.assertEquals(BigDecimal("1.12"), Value.roundToScale(BigDecimal("1.12"), 4, true))
    }
}
